File: relay/test/ztlp_relay/pipeline_test.exs
defmodule ZtlpRelay.PipelineTest do
  use ExUnit.Case

  alias ZtlpRelay.{Pipeline, Packet, Crypto, SessionRegistry, Stats}

  setup do
    # Reset stats before each test
    Stats.reset()
    :ok
  end

  describe "layer1_magic/1" do
    test "passes valid magic" do
      pkt = Packet.build_data(<<0::96>>, 0)
      raw = Packet.serialize(pkt)
      assert :ok = Pipeline.layer1_magic(raw)
    end

    test "drops invalid magic" do
      padding = :binary.copy(<<0>>, 40)
      assert {:drop, :invalid_magic} = Pipeline.layer1_magic(<<0xDE, 0xAD>> <> padding)
    end

    test "drops empty packet" do
      assert {:drop, :invalid_magic} = Pipeline.layer1_magic(<<>>)
    end
  end

  describe "layer2_session/1" do
    test "passes HELLO messages without session registration" do
      pkt = Packet.build_handshake(:hello, <<0::96>>)
      raw = Packet.serialize(pkt)
      assert :ok = Pipeline.layer2_session(raw)
    end

    test "passes HELLO_ACK messages without session registration" do
      pkt = Packet.build_handshake(:hello_ack, <<0::96>>)
      raw = Packet.serialize(pkt)
      assert :ok = Pipeline.layer2_session(raw)
    end

    test "drops data packets with unknown session" do
      session_id = :crypto.strong_rand_bytes(12)
      pkt = Packet.build_data(session_id, 0)
      raw = Packet.serialize(pkt)
      assert {:drop, :unknown_session} = Pipeline.layer2_session(raw)
    end

    test "passes data packets with registered session" do
      session_id = :crypto.strong_rand_bytes(12)
      peer_a = {{127, 0, 0, 1}, 5001}
      peer_b = {{127, 0, 0, 1}, 5002}
      SessionRegistry.register_session(session_id, peer_a, peer_b)

      pkt = Packet.build_data(session_id, 0)
      raw = Packet.serialize(pkt)
      assert :ok = Pipeline.layer2_session(raw)

      # Cleanup
      SessionRegistry.unregister_session(session_id)
    end

    test "passes handshake control packets with registered session" do
      session_id = :crypto.strong_rand_bytes(12)
      peer_a = {{127, 0, 0, 1}, 5001}
      peer_b = {{127, 0, 0, 1}, 5002}
      SessionRegistry.register_session(session_id, peer_a, peer_b)

      pkt = Packet.build_handshake(:rekey, session_id)
      raw = Packet.serialize(pkt)
      assert :ok = Pipeline.layer2_session(raw)

      SessionRegistry.unregister_session(session_id)
    end
  end

  describe "layer3_auth/2" do
    test "passes when session_key is nil (relay mode)" do
      pkt = Packet.build_data(<<0::96>>, 0)
      raw = Packet.serialize(pkt)
      assert :ok = Pipeline.layer3_auth(raw, nil)
    end

    test "passes HELLO messages regardless of key" do
      key = Crypto.generate_key()
      pkt = Packet.build_handshake(:hello, <<0::96>>)
      raw = Packet.serialize(pkt)
      assert :ok = Pipeline.layer3_auth(raw, key)
    end

    test "passes data packet with valid auth tag" do
      key = Crypto.generate_key()
      session_id = :crypto.strong_rand_bytes(12)

      # Build packet, compute auth tag over the AAD
      pkt = Packet.build_data(session_id, 1)
      raw_no_tag = Packet.serialize(pkt)
      {:ok, aad} = Packet.extract_aad(raw_no_tag)

      tag = Crypto.compute_header_auth_tag(key, aad)
      pkt_with_tag = %{pkt | header_auth_tag: tag}
      raw = Packet.serialize(pkt_with_tag)

      assert :ok = Pipeline.layer3_auth(raw, key)
    end

    test "drops data packet with invalid auth tag" do
      key = Crypto.generate_key()
      session_id = :crypto.strong_rand_bytes(12)

      pkt = Packet.build_data(session_id, 1, header_auth_tag: :crypto.strong_rand_bytes(16))
      raw = Packet.serialize(pkt)

      assert {:drop, :invalid_auth_tag} = Pipeline.layer3_auth(raw, key)
    end

    test "passes handshake packet with valid auth tag" do
      key = Crypto.generate_key()
      session_id = :crypto.strong_rand_bytes(12)

      pkt = Packet.build_handshake(:rekey, session_id)
      raw_no_tag = Packet.serialize(pkt)
      {:ok, aad} = Packet.extract_aad(raw_no_tag)

      tag = Crypto.compute_header_auth_tag(key, aad)
      pkt_with_tag = %{pkt | header_auth_tag: tag}
      raw = Packet.serialize(pkt_with_tag)

      assert :ok = Pipeline.layer3_auth(raw, key)
    end
  end

  describe "process/2 full pipeline" do
    test "drops non-ZTLP traffic at layer 1 and increments counter" do
      assert {:drop, 1, :invalid_magic} = Pipeline.process(<<0xDE, 0xAD>> <> :binary.copy(<<0>>, 40))
      stats = Stats.get_stats()
      assert stats.layer1_drops == 1
    end

    test "drops unknown session at layer 2 and increments counter" do
      session_id = :crypto.strong_rand_bytes(12)
      pkt = Packet.build_data(session_id, 0)
      raw = Packet.serialize(pkt)

      assert {:drop, 2, :unknown_session} = Pipeline.process(raw)
      stats = Stats.get_stats()
      assert stats.layer2_drops == 1
    end

    test "passes HELLO through all layers (relay mode, no key)" do
      pkt = Packet.build_handshake(:hello, <<0::96>>)
      raw = Packet.serialize(pkt)

      assert {:pass, parsed} = Pipeline.process(raw)
      assert parsed.msg_type == :hello
      stats = Stats.get_stats()
      assert stats.passed == 1
    end

    test "passes registered session data through all layers (relay mode)" do
      session_id = :crypto.strong_rand_bytes(12)
      peer_a = {{127, 0, 0, 1}, 5001}
      peer_b = {{127, 0, 0, 1}, 5002}
      SessionRegistry.register_session(session_id, peer_a, peer_b)

      pkt = Packet.build_data(session_id, 0)
      raw = Packet.serialize(pkt)

      assert {:pass, parsed} = Pipeline.process(raw)
      assert parsed.session_id == session_id
      stats = Stats.get_stats()
      assert stats.passed == 1

      SessionRegistry.unregister_session(session_id)
    end

    test "tracks multiple drops across layers" do
      # Layer 1 drops
      Pipeline.process(<<0xFF, 0xFF>>)
      Pipeline.process(<<0x00, 0x00>>)

      # Layer 2 drop
      pkt = Packet.build_data(:crypto.strong_rand_bytes(12), 0)
      Pipeline.process(Packet.serialize(pkt))

      stats = Stats.get_stats()
      assert stats.layer1_drops == 2
      assert stats.layer2_drops == 1
      assert stats.passed == 0
    end
  end
end

File: sdk/go/pipeline_test.go
package ztlp

import (
	"encoding/binary"
	"testing"
)

func TestComputeAndVerifyAuthTag(t *testing.T) {
	key := [32]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
		17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32}
	aad := []byte("some header aad bytes")

	tag := ComputeHeaderAuthTag(&key, aad)

	// Tag should not be zero
	allZero := true
	for _, b := range tag {
		if b != 0 {
			allZero = false
			break
		}
	}
	if allZero {
		t.Error("auth tag should not be all zeros")
	}

	// Verify should succeed with correct key and AAD
	if !VerifyHeaderAuthTag(&key, aad, tag[:]) {
		t.Error("verification should succeed with correct key and AAD")
	}

	// Verify should fail with wrong key
	wrongKey := [32]byte{0xFF}
	if VerifyHeaderAuthTag(&wrongKey, aad, tag[:]) {
		t.Error("verification should fail with wrong key")
	}

	// Verify should fail with wrong AAD
	if VerifyHeaderAuthTag(&key, []byte("wrong aad"), tag[:]) {
		t.Error("verification should fail with wrong AAD")
	}

	// Verify should fail with tampered tag
	tampered := make([]byte, AuthTagSize)
	copy(tampered, tag[:])
	tampered[0] ^= 0xFF
	if VerifyHeaderAuthTag(&key, aad, tampered) {
		t.Error("verification should fail with tampered tag")
	}
}

func TestComputeAuthTagDeterministic(t *testing.T) {
	key := [32]byte{0xAA}
	aad := []byte("deterministic test")

	tag1 := ComputeHeaderAuthTag(&key, aad)
	tag2 := ComputeHeaderAuthTag(&key, aad)

	if tag1 != tag2 {
		t.Error("same inputs should produce same tag")
	}
}

func TestLayer1MagicCheck(t *testing.T) {
	p := NewPipeline()

	// Valid magic
	validPacket := make([]byte, 4)
	binary.BigEndian.PutUint16(validPacket, Magic)
	if r := p.Layer1MagicCheck(validPacket); r != AdmissionPass {
		t.Errorf("valid magic: got %v, want Pass", r)
	}

	// Invalid magic
	invalidPacket := []byte{0xFF, 0xFF, 0x00, 0x00}
	if r := p.Layer1MagicCheck(invalidPacket); r != AdmissionDrop {
		t.Errorf("invalid magic: got %v, want Drop", r)
	}

	// Too short
	if r := p.Layer1MagicCheck([]byte{0x5A}); r != AdmissionDrop {
		t.Errorf("too short: got %v, want Drop", r)
	}

	// Empty
	if r := p.Layer1MagicCheck(nil); r != AdmissionDrop {
		t.Errorf("nil: got %v, want Drop", r)
	}
}

func TestLayer2SessionCheck(t *testing.T) {
	p := NewPipeline()

	// Register a session
	sid := SessionID{0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C}
	session := NewSessionState(sid, NodeID{}, [32]byte{}, [32]byte{}, false)
	p.RegisterSession(session)

	// Handshake HELLO should pass (any SessionID)
	helloHdr := NewHandshakeHeader(MsgTypeHello)
	helloData := helloHdr.Serialize()
	if r := p.Layer2SessionCheck(helloData); r != AdmissionPass {
		t.Errorf("HELLO: got %v, want Pass", r)
	}

	// Handshake HELLO_ACK should pass
	helloAckHdr := NewHandshakeHeader(MsgTypeHelloAck)
	helloAckData := helloAckHdr.Serialize()
	if r := p.Layer2SessionCheck(helloAckData); r != AdmissionPass {
		t.Errorf("HELLO_ACK: got %v, want Pass", r)
	}

	// Data packet with known SessionID should pass
	dataHdr := NewDataHeader(sid, 0)
	dataData := dataHdr.Serialize()
	if r := p.Layer2SessionCheck(dataData); r != AdmissionPass {
		t.Errorf("known session data: got %v, want Pass", r)
	}

	// Data packet with unknown SessionID should be dropped
	unknownSID := SessionID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
	unknownDataHdr := NewDataHeader(unknownSID, 0)
	unknownData := unknownDataHdr.Serialize()
	if r := p.Layer2SessionCheck(unknownData); r != AdmissionDrop {
		t.Errorf("unknown session data: got %v, want Drop", r)
	}

	// Handshake with known session and non-HELLO type should pass
	rekeyHdr := NewHandshakeHeader(MsgTypeRekey)
	rekeyHdr.SessionID = sid
	rekeyData := rekeyHdr.Serialize()
	if r := p.Layer2SessionCheck(rekeyData); r != AdmissionPass {
		t.Errorf("known session rekey: got %v, want Pass", r)
	}

	// Handshake with unknown session and non-HELLO type should drop
	rekeyHdr2 := NewHandshakeHeader(MsgTypeRekey)
	rekeyHdr2.SessionID = unknownSID
	rekeyData2 := rekeyHdr2.Serialize()
	if r := p.Layer2SessionCheck(rekeyData2); r != AdmissionDrop {
		t.Errorf("unknown session rekey: got %v, want Drop", r)
	}
}

func TestLayer3AuthCheck(t *testing.T) {
	p := NewPipeline()

	// Set up a session with known keys
	sid := SessionID{0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C}
	sendKey := [32]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
		17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32}
	recvKey := sendKey // For simplicity, same key both directions
	session := NewSessionState(sid, NodeID{}, sendKey, recvKey, false)
	p.RegisterSession(session)

	// HELLO should pass without auth check
	helloHdr := NewHandshakeHeader(MsgTypeHello)
	helloData := helloHdr.Serialize()
	if r := p.Layer3AuthCheck(helloData); r != AdmissionPass {
		t.Errorf("HELLO: got %v, want Pass", r)
	}

	// Data packet with valid auth tag should pass
	dataHdr := NewDataHeader(sid, 42)
	aad := dataHdr.AADBytes()
	dataHdr.HeaderAuthTag = ComputeHeaderAuthTag(&sendKey, aad)
	dataData := dataHdr.Serialize()
	if r := p.Layer3AuthCheck(dataData); r != AdmissionPass {
		t.Errorf("valid auth tag: got %v, want Pass", r)
	}

	// Data packet with invalid auth tag should drop
	badHdr := NewDataHeader(sid, 43)
	badHdr.HeaderAuthTag = [AuthTagSize]byte{0xDE, 0xAD}
	badData := badHdr.Serialize()
	if r := p.Layer3AuthCheck(badData); r != AdmissionDrop {
		t.Errorf("invalid auth tag: got %v, want Drop", r)
	}
}

func TestPipelineProcess(t *testing.T) {
	p := NewPipeline()

	// Non-ZTLP packet should be dropped at Layer 1
	if r := p.Process([]byte{0x00, 0x00, 0x00, 0x00}); r != AdmissionDrop {
		t.Errorf("bad magic: got %v, want Drop", r)
	}
	snap := p.Counters.Snapshot()
	if snap.Layer1Drops != 1 {
		t.Errorf("layer1 drops: %d, want 1", snap.Layer1Drops)
	}

	// HELLO should pass all layers (no session required, no auth check)
	helloHdr := NewHandshakeHeader(MsgTypeHello)
	helloData := helloHdr.Serialize()
	if r := p.Process(helloData); r != AdmissionPass {
		t.Errorf("HELLO: got %v, want Pass", r)
	}
	snap = p.Counters.Snapshot()
	if snap.Passed != 1 {
		t.Errorf("passed: %d, want 1", snap.Passed)
	}

	// Data with unknown session should be dropped at Layer 2
	unknownDataHdr := NewDataHeader(SessionID{0xFF}, 0)
	unknownData := unknownDataHdr.Serialize()
	if r := p.Process(unknownData); r != AdmissionDrop {
		t.Errorf("unknown session: got %v, want Drop", r)
	}
	snap = p.Counters.Snapshot()
	if snap.Layer2Drops != 1 {
		t.Errorf("layer2 drops: %d, want 1", snap.Layer2Drops)
	}
}

func TestPipelineRegisterRemove(t *testing.T) {
	p := NewPipeline()
	sid := SessionID{1, 2, 3}
	session := NewSessionState(sid, NodeID{}, [32]byte{}, [32]byte{}, false)

	// Register
	p.RegisterSession(session)
	if s := p.GetSession(sid); s == nil {
		t.Error("session should be registered")
	}

	// Remove
	p.RemoveSession(sid)
	if s := p.GetSession(sid); s != nil {
		t.Error("session should be removed")
	}
}

func TestAdmissionResultString(t *testing.T) {
	tests := []struct {
		r    AdmissionResult
		want string
	}{
		{AdmissionPass, "Pass"},
		{AdmissionDrop, "Drop"},
		{AdmissionRateLimit, "RateLimit"},
		{AdmissionResult(99), "Unknown"},
	}
	for _, tt := range tests {
		if got := tt.r.String(); got != tt.want {
			t.Errorf("%d.String() = %q, want %q", tt.r, got, tt.want)
		}
	}
}
